File: Data Structures & Algorithms/valid-sudoku/submission-2.cs
public class Solution {
    public bool IsValidSudoku(char[][] board) {
        var rows = new Dictionary<int,HashSet<char>>();
        var cols = new Dictionary<int,HashSet<char>>();
        var boxes = new Dictionary<string,HashSet<char>>();
        for(int i=0;i<board.Length;i++){
            for(int j=0;j<board[0].Length;j++){
                if(board[i][j]=='.') continue;
                if((rows.ContainsKey(i) && rows[i].Contains(board[i][j]))||
                (cols.ContainsKey(j) && cols[j].Contains(board[i][j]))||
                (boxes.ContainsKey((i/3)+"-"+(j/3)) && boxes[(i/3)+"-"+(j/3)].Contains(board[i][j]))){
                    return false;
                }
                if(!rows.ContainsKey(i)) rows[i] = new HashSet<char>();
                if(!cols.ContainsKey(j)) cols[j] = new HashSet<char>();
                if(!boxes.ContainsKey((i/3)+"-"+(j/3))) boxes[(i/3)+"-"+(j/3)] = new HashSet<char>();
                rows[i].Add(board[i][j]);
                cols[j].Add(board[i][j]);
                boxes[(i/3)+"-"+(j/3)].Add(board[i][j]);
            }
        }
        return true;
    }
}

File: Data Structures & Algorithms/n-queens/submission-2.java
class Solution {

    Set<Integer> col = new HashSet<>();
    Set<Integer> posDiag = new HashSet<>();
    Set<Integer> negDiag = new HashSet<>();

    public List<List<String>> solveNQueens(int n) {
        List<List<String>> res = new ArrayList<>();
        char board[][] = new char[n][n];
        for(int i=0;i<n;i++){
            for(int j=0;j<n;j++){
                board[i][j] = '.';
            }
        }
        backtrack(0,board,res);
        return res;
    }

    public void backtrack(int r, char[][] board, List<List<String>> res){
        if(r==board.length){
            List<String> ls = new ArrayList<>();
            for(int i=0;i<board.length;i++){
                StringBuilder sb = new StringBuilder();
                for(int j=0;j<board[0].length;j++){
                    sb.append(board[i][j]);
                }
                ls.add(sb.toString());
            }
            res.add(ls);
            return;
        }
        for(int c=0;c<board[0].length;c++){
            if(col.contains(c) || posDiag.contains(r+c) || negDiag.contains(r-c)) continue;
            col.add(c);
            posDiag.add(r+c);
            negDiag.add(r-c);
            board[r][c] = 'Q';
            backtrack(r+1, board, res);
            col.remove(c);
            posDiag.remove(r+c);
            negDiag.remove(r-c);
            board[r][c] = '.';
        }
    }
}
